core/controllers/email_dashboard_test.py

```diff
@@ -117,3 +118,300 @@ def test_that_exception_is_raised_for_invalid_input(self):
+            self.process_and_flush_pending_tasks()
+            # Check that qualified users are valid.
+            query_models = user_models.UserQueryModel.query().fetch()
+            self.assertEqual(len(query_models[0].user_ids), 2)
+
+            # Send email from email dashboard result page.
+            self.login(self.SUBMITTER_EMAIL)
+            csrf_token = self.get_csrf_token_from_response(
+                self.testapp.get(
+                    '/emaildashboardresult/%s' % query_models[0].id))
+            self.post_json(
+                '/emaildashboardresult/%s' % query_models[0].id, {
+                    'data': {
+                        'email_subject': 'subject',
+                        'email_body': 'body',
+                        'email_option': 'custom',
+                        'num_of_users_to_send': 1,
+                        'email_intent': 'bulk_email_marketing'
+                    }}, csrf_token)
+            self.logout()
+
+            # Check that emails are sent to max n qualified users.
+            # One email is sent to submitter for query completion and second
+            # is sent to one of the 2 qualified users.
+            messages = self.mail_stub.get_sent_messages()
+            self.assertEqual(len(messages), 2)
+
+    def test_that_no_emails_are_sent_if_query_is_canceled(self):
+        self.login(self.SUBMITTER_EMAIL)
+        csrf_token = self.get_csrf_token_from_response(
+            self.testapp.get('/emaildashboard'))
+        self.post_json(
+            '/emaildashboarddatahandler', {
+                'data': {
+                    'has_not_logged_in_for_n_days': None,
+                    'inactive_in_last_n_days': None,
+                    'created_at_least_n_exps': 1,
+                    'created_fewer_than_n_exps': None,
+                    'edited_at_least_n_exps': None,
+                    'edited_fewer_than_n_exps': None
+                }}, csrf_token)
+        self.logout()
+
+        query_models = user_models.UserQueryModel.query().fetch()
+
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True):
+            self.process_and_flush_pending_tasks()
+            # Check that qualified users are valid.
+            query_models = user_models.UserQueryModel.query().fetch()
+            self.assertEqual(len(query_models[0].user_ids), 2)
+
+            # Send email from email dashboard result page.
+            self.login(self.SUBMITTER_EMAIL)
+            csrf_token = self.get_csrf_token_from_response(
+                self.testapp.get(
+                    '/emaildashboardresult/%s' % query_models[0].id))
+            self.post_json(
+                '/emaildashboardresult/%s' % query_models[0].id, {
+                    'data': {
+                        'email_subject': 'subject',
+                        'email_body': 'body',
+                        'email_option': 'cancel',
+                        'num_of_users_to_send': None,
```

Comment:
Erh -- this is ugly. I think you should have separate handlers in the controller for cancelling and sending.